core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -650,6 +653,28 @@ else if (redistributeWrites) {
                     partitioningScheme = Optional.of(new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), newSource.getNode().getOutputSymbols()));
                 }
             }
+            else if (scaleWriters
+                    && writerTarget.supportsReportingWrittenBytes(plannerContext.getMetadata(), session)
+                    && writerTarget.supportsMultipleWritersPerPartition(plannerContext.getMetadata(), session)
+                    // do not insert an exchange if partitioning is compatible
+                    && !newSource.getProperties().isCompatibleTablePartitioningWith(partitioningScheme.get().getPartitioning(), false, plannerContext.getMetadata(), session)) {
+                if (partitioningScheme.get().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+                    partitioningScheme = Optional.of(partitioningScheme.get().withPartitioningHandle(SCALED_WRITER_HASH_DISTRIBUTION));
+                }
+                else {
+                    PartitioningHandle partitioningHandle = partitioningScheme.get().getPartitioning().getHandle();
+                    verify(!(partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle));
+                    verify(
+                            partitioningScheme.get().getPartitioning().getArguments().stream().noneMatch(Partitioning.ArgumentBinding::isConstant),
+                            "Table writer partitioning has constant arguments");
+                    partitioningScheme = Optional.of(partitioningScheme.get().withPartitioningHandle(
+                            new PartitioningHandle(
+                                    partitioningHandle.getCatalogHandle(),
+                                    partitioningHandle.getTransactionHandle(),
+                                    partitioningHandle.getConnectorHandle(),
+                                    true)));
```

Comment:
can we always do that? Doesn't it depend on nature of connector partitioning?

Reply:
We are checking for `&& writerTarget.supportsMultipleWritersPerPartition(plannerContext.getMetadata(), session)`